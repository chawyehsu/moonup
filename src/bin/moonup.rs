#[tokio::main]
pub async fn main() {
    if let Err(err) = moonup::cli::start().await {
        eprintln!("{err:?}");
        std::process::exit(1);
    }
}
