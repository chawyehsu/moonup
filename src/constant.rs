/// The default Moon home directory is `~/.moon`
pub const MOON_DIR: &str = ".moon";

/// The default MoonUp home directory is `~/.moonup`
pub const MOONUP_DIR: &str = ".moonup";

/// The URL of the MoonBit release index
pub const INDEX_URL: &str = "https://chawyehsu.github.io/moonbit-binaries/index.json";

/// The timeout for reading HTTP responses, in hours
pub const INDEX_EXPIRATION: i64 = 2;

/// The timeout for reading HTTP responses, in seconds
pub const HTTP_READ_TIMEOUT: u64 = 5 * 60;
