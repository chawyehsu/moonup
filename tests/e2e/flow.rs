use insta::glob;
use insta_cmd::assert_cmd_snapshot;
use mockito::Server;
use moonup::constant;
use std::{env, fs};

use crate::util::{self, TestWorkspace};

#[test]
fn test_basic_flow() {
    util::apply_common_filters!();

    let ws = TestWorkspace::new();

    // No toolchain installed
    assert_cmd_snapshot!("moonup_show", ws.cli().arg("show"));

    // Set default toolchain
    assert_cmd_snapshot!("moonup_default", ws.cli().arg("default").arg("latest"));

    // Pin toolchain
    let project_path = ws.project_path();
    fs::create_dir_all(project_path).expect("should create project directory");
    env::set_current_dir(project_path).expect("should set current directory");

    assert_cmd_snapshot!("moonup_pin", ws.cli().arg("pin").arg("nightly"));
    assert!(project_path.join(constant::TOOLCHAIN_FILE).exists());

    env::set_current_dir(ws.tempdir()).expect("should restore current directory");
}

#[test]
fn test_flow_with_network_mock() {
    util::apply_common_filters!();

    let ws = TestWorkspace::new();
    let mut cli = ws.cli();

    let mut s = Server::new();

    // setup mock server for dist_server
    // NOTE(chawyehsu): insta glob! macro does not support using one single path
    // (for example, `glob!("../fixtures/dist_server/**/*.json", |path: &Path| { ... })`)
    // to match files outside the current directory, a base path is required.
    glob!("../fixtures/dist_server", "**/*.json", |path: &Path| {
        let path = path.display().to_string();
        #[cfg(target_os = "windows")]
        let path = path.replace("\\", "/");

        let pathname = path.rsplit_once("dist_server").unwrap().1;
        // println!("Mocking: {} (fullpath: {})", pathname, path);

        s.mock("GET", pathname)
            .with_body_from_file(path)
            .with_header("content-type", "application/json")
            .create();
    });

    assert_cmd_snapshot!(cli
        .env(constant::ENVNAME_MOONUP_DIST_SERVER, s.url())
        .arg("install")
        .arg("--list-available"));
}

/// Test flow with production networking
#[cfg(feature = "test-liveinstall")]
mod liveinstall {
    use super::*;

    #[test]
    fn test_flow_with_network() {
        util::apply_common_filters!();

        let ws = TestWorkspace::new();
        let test_install_version = "0.1.20241231+ba15a9a4e";

        // Test self update (should be no update)
        assert_cmd_snapshot!("moonup_selfupdate", ws.cli().arg("self-update"));

        // Install a specific version of the toolchain
        assert_cmd_snapshot!("install", ws.cli().arg("install").arg(test_install_version));

        // List installed toolchains, installed toolchain should be listed
        assert_cmd_snapshot!("moonup_list_1", ws.cli().arg("list"));

        // Pin toolchain
        let project_path = ws.project_path();
        let pin_file = project_path.join(constant::TOOLCHAIN_FILE);

        fs::create_dir_all(project_path).expect("should create project directory");
        env::set_current_dir(project_path).expect("should set current directory");

        assert_cmd_snapshot!("moonup_pin", ws.cli().arg("pin").arg(test_install_version));
        assert!(pin_file.exists());

        let moon_path = ws.moon_home().display().to_string();
        let current_path = env::var("PATH").unwrap_or_else(|_| String::new());
        let updated_path = if current_path.is_empty() {
            moon_path
        } else {
            let env_separator = if cfg!(windows) { ";" } else { ":" };
            format!("{}{}{}", moon_path, env_separator, current_path)
        };
        // println!("Updated PATH: {}", updated_path);

        temp_env::with_var("PATH", Some(updated_path.clone()), || {
            let mut cmd_moon = ws.cmd(std::ffi::OsStr::new("moon"));
            assert_cmd_snapshot!("use_pinned_version", cmd_moon.arg("version").arg("--all"));
        });

        // Remove the pinned toolchain
        std::fs::remove_file(pin_file).expect("should remove pinned toolchain file");

        // Set default toolchain
        assert_cmd_snapshot!(
            "moonup_default",
            ws.cli().arg("default").arg(test_install_version)
        );

        temp_env::with_var("PATH", Some(updated_path), || {
            let mut cmd_moon = ws.cmd(std::ffi::OsStr::new("moon"));
            assert_cmd_snapshot!("use_default_version", cmd_moon.arg("version"));

            // Uninstall the installed toolchain
            assert_cmd_snapshot!(
                "moonup_uninstall",
                ws.cli().arg("uninstall").arg(test_install_version)
            );

            // Toolchain should be uninstalled, cache should be removed
            let install_path = ws
                .moonup_home()
                .join("toolchains")
                .join(test_install_version);
            let cache_path = ws
                .moonup_home()
                .join("downloads")
                .join(test_install_version);

            assert!(!install_path.exists());
            assert!(!cache_path.exists());

            // List installed toolchains, no toolchain should be listed
            assert_cmd_snapshot!("moonup_list_2", ws.cli().arg("list"));
        });

        env::set_current_dir(ws.tempdir()).expect("should restore current directory");
    }
}
