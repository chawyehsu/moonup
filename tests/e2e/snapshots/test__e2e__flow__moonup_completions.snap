---
source: tests/e2e/flow.rs
info:
  program: moonup
  args:
    - completions
    - zsh
---
success: true
exit_code: 0
----- stdout -----
#compdef moonup

autoload -U is-at-least

_moonup() {
    typeset -A opt_args
    typeset -a _arguments_options
    local ret=1

    if is-at-least 5.2; then
        _arguments_options=(-s -S -C)
    else
        _arguments_options=(-s -C)
    fi

    local context curcontext="$curcontext" state line
    _arguments "${_arguments_options[@]}" : \
'*-v[Increase logging verbosity]' \
'*--verbose[Increase logging verbosity]' \
'(-v --verbose)*-q[Decrease logging verbosity]' \
'(-v --verbose)*--quiet[Decrease logging verbosity]' \
'-h[Print help]' \
'--help[Print help]' \
'-V[Print version]' \
'--version[Print version]' \
":: :_moonup_commands" \
"*::: :->moonup" \
&& ret=0
    case $state in
    (moonup)
        words=($line[1] "${words[@]}")
        (( CURRENT += 1 ))
        curcontext="${curcontext%:*:*}:moonup-command-$line[1]:"
        case $line[1] in
            (completions)
_arguments "${_arguments_options[@]}" : \
'*-v[Increase logging verbosity]' \
'*--verbose[Increase logging verbosity]' \
'(-v --verbose)*-q[Decrease logging verbosity]' \
'(-v --verbose)*--quiet[Decrease logging verbosity]' \
'-h[Print help]' \
'--help[Print help]' \
':shell -- The shell type:(bash elvish fish powershell zsh)' \
&& ret=0
;;
(default)
_arguments "${_arguments_options[@]}" : \
'*-v[Increase logging verbosity]' \
'*--verbose[Increase logging verbosity]' \
'(-v --verbose)*-q[Decrease logging verbosity]' \
'(-v --verbose)*--quiet[Decrease logging verbosity]' \
'-h[Print help]' \
'--help[Print help]' \
'::toolchain -- Toolchain version tag or channel name \[latest, nightly, bleeding\]:_default' \
&& ret=0
;;
(install)
_arguments "${_arguments_options[@]}" : \
'--list-available[List available toolchains]' \
'*-v[Increase logging verbosity]' \
'*--verbose[Increase logging verbosity]' \
'(-v --verbose)*-q[Decrease logging verbosity]' \
'(-v --verbose)*--quiet[Decrease logging verbosity]' \
'-h[Print help]' \
'--help[Print help]' \
'::toolchain -- Toolchain version tag or channel name \[latest, nightly, bleeding\]:_default' \
&& ret=0
;;
(i)
_arguments "${_arguments_options[@]}" : \
'--list-available[List available toolchains]' \
'*-v[Increase logging verbosity]' \
'*--verbose[Increase logging verbosity]' \
'(-v --verbose)*-q[Decrease logging verbosity]' \
'(-v --verbose)*--quiet[Decrease logging verbosity]' \
'-h[Print help]' \
'--help[Print help]' \
'::toolchain -- Toolchain version tag or channel name \[latest, nightly, bleeding\]:_default' \
&& ret=0
;;
(list)
_arguments "${_arguments_options[@]}" : \
'*-v[Increase logging verbosity]' \
'*--verbose[Increase logging verbosity]' \
'(-v --verbose)*-q[Decrease logging verbosity]' \
'(-v --verbose)*--quiet[Decrease logging verbosity]' \
'-h[Print help]' \
'--help[Print help]' \
&& ret=0
;;
(ls)
_arguments "${_arguments_options[@]}" : \
'*-v[Increase logging verbosity]' \
'*--verbose[Increase logging verbosity]' \
'(-v --verbose)*-q[Decrease logging verbosity]' \
'(-v --verbose)*--quiet[Decrease logging verbosity]' \
'-h[Print help]' \
'--help[Print help]' \
&& ret=0
;;
(pin)
_arguments "${_arguments_options[@]}" : \
'*-v[Increase logging verbosity]' \
'*--verbose[Increase logging verbosity]' \
'(-v --verbose)*-q[Decrease logging verbosity]' \
'(-v --verbose)*--quiet[Decrease logging verbosity]' \
'-h[Print help]' \
'--help[Print help]' \
'::toolchain -- Toolchain version tag or channel name \[latest, nightly, bleeding\]:_default' \
&& ret=0
;;
(run)
_arguments "${_arguments_options[@]}" : \
'*-v[Increase logging verbosity]' \
'*--verbose[Increase logging verbosity]' \
'(-v --verbose)*-q[Decrease logging verbosity]' \
'(-v --verbose)*--quiet[Decrease logging verbosity]' \
'-h[Print help]' \
'--help[Print help]' \
':toolchain -- The toolchain to use for running the command:_default' \
'*::command -- The command to run, with arguments if any:_cmdambivalent' \
&& ret=0
;;
(x)
_arguments "${_arguments_options[@]}" : \
'*-v[Increase logging verbosity]' \
'*--verbose[Increase logging verbosity]' \
'(-v --verbose)*-q[Decrease logging verbosity]' \
'(-v --verbose)*--quiet[Decrease logging verbosity]' \
'-h[Print help]' \
'--help[Print help]' \
':toolchain -- The toolchain to use for running the command:_default' \
'*::command -- The command to run, with arguments if any:_cmdambivalent' \
&& ret=0
;;
(self-update)
_arguments "${_arguments_options[@]}" : \
'*-v[Increase logging verbosity]' \
'*--verbose[Increase logging verbosity]' \
'(-v --verbose)*-q[Decrease logging verbosity]' \
'(-v --verbose)*--quiet[Decrease logging verbosity]' \
'-h[Print help]' \
'--help[Print help]' \
&& ret=0
;;
(uninstall)
_arguments "${_arguments_options[@]}" : \
'--clear[Invalidate and remove all cached downloads]' \
'--keep-cache[Keep the cached downloads of the toolchain]' \
'*-v[Increase logging verbosity]' \
'*--verbose[Increase logging verbosity]' \
'(-v --verbose)*-q[Decrease logging verbosity]' \
'(-v --verbose)*--quiet[Decrease logging verbosity]' \
'-h[Print help (see more with '\''--help'\'')]' \
'--help[Print help (see more with '\''--help'\'')]' \
'*::toolchain -- The toolchain(s) to uninstall:_default' \
&& ret=0
;;
(update)
_arguments "${_arguments_options[@]}" : \
'*-v[Increase logging verbosity]' \
'*--verbose[Increase logging verbosity]' \
'(-v --verbose)*-q[Decrease logging verbosity]' \
'(-v --verbose)*--quiet[Decrease logging verbosity]' \
'-h[Print help]' \
'--help[Print help]' \
&& ret=0
;;
(u)
_arguments "${_arguments_options[@]}" : \
'*-v[Increase logging verbosity]' \
'*--verbose[Increase logging verbosity]' \
'(-v --verbose)*-q[Decrease logging verbosity]' \
'(-v --verbose)*--quiet[Decrease logging verbosity]' \
'-h[Print help]' \
'--help[Print help]' \
&& ret=0
;;
(which)
_arguments "${_arguments_options[@]}" : \
'*-v[Increase logging verbosity]' \
'*--verbose[Increase logging verbosity]' \
'(-v --verbose)*-q[Decrease logging verbosity]' \
'(-v --verbose)*--quiet[Decrease logging verbosity]' \
'-h[Print help]' \
'--help[Print help]' \
':command -- The command to inspect:_default' \
&& ret=0
;;
(help)
_arguments "${_arguments_options[@]}" : \
":: :_moonup__help_commands" \
"*::: :->help" \
&& ret=0

    case $state in
    (help)
        words=($line[1] "${words[@]}")
        (( CURRENT += 1 ))
        curcontext="${curcontext%:*:*}:moonup-help-command-$line[1]:"
        case $line[1] in
            (completions)
_arguments "${_arguments_options[@]}" : \
&& ret=0
;;
(default)
_arguments "${_arguments_options[@]}" : \
&& ret=0
;;
(install)
_arguments "${_arguments_options[@]}" : \
&& ret=0
;;
(list)
_arguments "${_arguments_options[@]}" : \
&& ret=0
;;
(pin)
_arguments "${_arguments_options[@]}" : \
&& ret=0
;;
(run)
_arguments "${_arguments_options[@]}" : \
&& ret=0
;;
(self-update)
_arguments "${_arguments_options[@]}" : \
&& ret=0
;;
(uninstall)
_arguments "${_arguments_options[@]}" : \
&& ret=0
;;
(update)
_arguments "${_arguments_options[@]}" : \
&& ret=0
;;
(which)
_arguments "${_arguments_options[@]}" : \
&& ret=0
;;
(help)
_arguments "${_arguments_options[@]}" : \
&& ret=0
;;
        esac
    ;;
esac
;;
        esac
    ;;
esac
}

(( $+functions[_moonup_commands] )) ||
_moonup_commands() {
    local commands; commands=(
'completions:Generate shell completions' \
'default:Set the default toolchain' \
'install:Install or update a MoonBit toolchain' \
'i:Install or update a MoonBit toolchain' \
'list:List installed and active toolchains' \
'ls:List installed and active toolchains' \
'pin:Pin the MoonBit toolchain to a specific version' \
'run:Run a command with a specific toolchain' \
'x:Run a command with a specific toolchain' \
'self-update:Update Moonup to the latest version' \
'uninstall:Uninstall a MoonBit toolchain' \
'update:Update MoonBit toolchains' \
'u:Update MoonBit toolchains' \
'which:Show the actual binary that will be run for a given command' \
'help:Print this message or the help of the given subcommand(s)' \
    )
    _describe -t commands 'moonup commands' commands "$@"
}
(( $+functions[_moonup__completions_commands] )) ||
_moonup__completions_commands() {
    local commands; commands=()
    _describe -t commands 'moonup completions commands' commands "$@"
}
(( $+functions[_moonup__default_commands] )) ||
_moonup__default_commands() {
    local commands; commands=()
    _describe -t commands 'moonup default commands' commands "$@"
}
(( $+functions[_moonup__help_commands] )) ||
_moonup__help_commands() {
    local commands; commands=(
'completions:Generate shell completions' \
'default:Set the default toolchain' \
'install:Install or update a MoonBit toolchain' \
'list:List installed and active toolchains' \
'pin:Pin the MoonBit toolchain to a specific version' \
'run:Run a command with a specific toolchain' \
'self-update:Update Moonup to the latest version' \
'uninstall:Uninstall a MoonBit toolchain' \
'update:Update MoonBit toolchains' \
'which:Show the actual binary that will be run for a given command' \
'help:Print this message or the help of the given subcommand(s)' \
    )
    _describe -t commands 'moonup help commands' commands "$@"
}
(( $+functions[_moonup__help__completions_commands] )) ||
_moonup__help__completions_commands() {
    local commands; commands=()
    _describe -t commands 'moonup help completions commands' commands "$@"
}
(( $+functions[_moonup__help__default_commands] )) ||
_moonup__help__default_commands() {
    local commands; commands=()
    _describe -t commands 'moonup help default commands' commands "$@"
}
(( $+functions[_moonup__help__help_commands] )) ||
_moonup__help__help_commands() {
    local commands; commands=()
    _describe -t commands 'moonup help help commands' commands "$@"
}
(( $+functions[_moonup__help__install_commands] )) ||
_moonup__help__install_commands() {
    local commands; commands=()
    _describe -t commands 'moonup help install commands' commands "$@"
}
(( $+functions[_moonup__help__list_commands] )) ||
_moonup__help__list_commands() {
    local commands; commands=()
    _describe -t commands 'moonup help list commands' commands "$@"
}
(( $+functions[_moonup__help__pin_commands] )) ||
_moonup__help__pin_commands() {
    local commands; commands=()
    _describe -t commands 'moonup help pin commands' commands "$@"
}
(( $+functions[_moonup__help__run_commands] )) ||
_moonup__help__run_commands() {
    local commands; commands=()
    _describe -t commands 'moonup help run commands' commands "$@"
}
(( $+functions[_moonup__help__self-update_commands] )) ||
_moonup__help__self-update_commands() {
    local commands; commands=()
    _describe -t commands 'moonup help self-update commands' commands "$@"
}
(( $+functions[_moonup__help__uninstall_commands] )) ||
_moonup__help__uninstall_commands() {
    local commands; commands=()
    _describe -t commands 'moonup help uninstall commands' commands "$@"
}
(( $+functions[_moonup__help__update_commands] )) ||
_moonup__help__update_commands() {
    local commands; commands=()
    _describe -t commands 'moonup help update commands' commands "$@"
}
(( $+functions[_moonup__help__which_commands] )) ||
_moonup__help__which_commands() {
    local commands; commands=()
    _describe -t commands 'moonup help which commands' commands "$@"
}
(( $+functions[_moonup__install_commands] )) ||
_moonup__install_commands() {
    local commands; commands=()
    _describe -t commands 'moonup install commands' commands "$@"
}
(( $+functions[_moonup__list_commands] )) ||
_moonup__list_commands() {
    local commands; commands=()
    _describe -t commands 'moonup list commands' commands "$@"
}
(( $+functions[_moonup__pin_commands] )) ||
_moonup__pin_commands() {
    local commands; commands=()
    _describe -t commands 'moonup pin commands' commands "$@"
}
(( $+functions[_moonup__run_commands] )) ||
_moonup__run_commands() {
    local commands; commands=()
    _describe -t commands 'moonup run commands' commands "$@"
}
(( $+functions[_moonup__self-update_commands] )) ||
_moonup__self-update_commands() {
    local commands; commands=()
    _describe -t commands 'moonup self-update commands' commands "$@"
}
(( $+functions[_moonup__uninstall_commands] )) ||
_moonup__uninstall_commands() {
    local commands; commands=()
    _describe -t commands 'moonup uninstall commands' commands "$@"
}
(( $+functions[_moonup__update_commands] )) ||
_moonup__update_commands() {
    local commands; commands=()
    _describe -t commands 'moonup update commands' commands "$@"
}
(( $+functions[_moonup__which_commands] )) ||
_moonup__which_commands() {
    local commands; commands=()
    _describe -t commands 'moonup which commands' commands "$@"
}

if [ "$funcstack[1]" = "_moonup" ]; then
    _moonup "$@"
else
    compdef _moonup moonup
fi


----- stderr -----
